probe-rs/src/bin/probe-rs/cmd/run/test_run_mode.rs

```diff
@@ -0,0 +1,292 @@
+        // Then we wait until the target invokes SYS_EXIT (0x18) or SYS_EXIT_EXTENDED(0x20) with the exit code
+        let halt_handler = |halt_reason: HaltReason, core: &mut Core| {
+            match halt_reason {
+                HaltReason::Breakpoint(BreakpointCause::Semihosting(cmd)) => match cmd {
+                    SemihostingCommand::GetCommandLine(request) if !cmdline_requested => {
+                        let cmdline = format!("run {}", test.name);
+                        tracing::info!("target asked for cmdline. send '{}'", cmdline.as_str());
+                        cmdline_requested = true;
+                        request.write_command_line_to_target(core, cmdline.as_str())?; //TODO: fix default retreg if this is not called
+                        Ok(None) // Continue running
+                    }
+                    SemihostingCommand::ExitSuccess if cmdline_requested => Ok(Some(true)),
+                    SemihostingCommand::ExitError(_) if cmdline_requested => Ok(Some(false)),
+                    other => {
+                        // Invalid sequence of semihosting calls => Abort testing altogether
+                        Err(anyhow!(
+                            "Unexpected semihosting command {:?} cmdline_requested: {:?}",
+                            other,
+                            cmdline_requested
+                        ))
+                    }
+                },
+                e => {
+                    // Exception occurred (e.g. hardfault) => Abort testing altogether
+                    Err(anyhow!("The CPU halted unexpectedly: {:?}. Test should signal failure via a panic handler that calls `semihosting::proces::abort()` instead", e))
+                }
+            }
+        };
+
+        match session_and_runloop
+            .run_loop
+            .run_until(core, true, true, Some(timeout), halt_handler)
+        {
+            Ok(ReturnReason::Timeout) => {
+                Err(Failed::from(format!("Test timed out after {:?}", timeout)))
+            }
+            Ok(ReturnReason::User) => {
+                eprintln!("Test {} was aborted by the user with CTRL + C", test.name);
+                // We do not mark the test as failed and instead exit the process
+                std::process::exit(1);
+            }
+            Ok(ReturnReason::Predicate(exit_status)) => {
+                let should_exit_successfully = !test.should_panic;
+                if exit_status == should_exit_successfully {
+                    Ok(())
+                } else {
+                    if !exit_status {
+                        print_stacktrace(
+                            core,
+                            Path::new(session_and_runloop.run_loop.path.as_str()),
+                        )?;
+                    }
+                    Err(Failed::from(format!(
+                        "Test should have {} but it {}",
+                        if test.should_panic {
```

Comment:
This code feels like both `test.should_panic` and `exit_status` should be an `enum TestStatus { Passed, Panicked }`.

Reply:
Good idea! fixed.